website/blog/2024-04-25-scaling-data-pipelines-kuda.mdx

```diff
@@ -0,0 +1,254 @@
+description: "How Kuda leveraged dbt incremental models to reduce costs, speed up pipelines, and scale confidently."
+slug: scaling-data-pipelines-fintech
+authors: [adedamola_onabanjo]
+tags: [analytics, dbt Cloud, fintech, BigQuery]
+hide_table_of_contents: false
+date: 2025-04-23
+is_featured: true
+---
+
+## Introduction
+
+Building scalable data pipelines in a fast-growing fintech can feel like fixing a bike while riding it. You must keep insights flowing even as data volumes explode. At Kuda (a Nigerian neo-bank), we faced this problem as our user base surged. Traditional batch ETL (rebuilding entire tables each run) started to buckle; pipelines took hours, and costs ballooned. We needed to keep data fresh without reprocessing everything. Our solution was to leverage dbt’s incremental models, which process only new or changed records. This dramatically cut run times and curbed our BigQuery costs, letting us scale efficiently.
+
+<!-- truncate -->
+
+## Challenges in Scaling
+
+Rapid growth brought some serious scaling challenges, and the most important were:
+
+- **Performance**: Our nightly full-refresh models that once took minutes began taking hours as data grew. For example, our core transactions table became too slow to rebuild from scratch for each update. Analytics dashboards lagged, and stakeholders lost timely insights. In real-time fintech, such latency is unacceptable.
+
+- **Cost**: More data and longer processing also drove up our BigQuery bills. Scanning a 2TB table every hour to grab a few MB of new data was wasteful. Under BigQuery’s on-demand pricing model, this could rack up thousands of dollars per month. We needed to increase throughput without scaling cost linearly, which meant rethinking our processing to avoid full table scans.
+
+- **Data** Integrity: As pipelines and dependencies multiplied, so did the risk of inconsistencies or failures. Any new approach had to maintain accuracy and consistency even as we sped things up.
+
+## Approach: Incremental Models & Key Strategies
+
+We tackled these issues by embracing dbt’s incremental models, which process only new or updated records since the last run. Instead of monolithic daily rebuilds, our models continuously ingested changes in small bites. Below, we outline our key incremental strategies—`append`, `insert_overwrite`, and `merge`—and how we tuned performance and cost.
+
+### Append Strategy
+
+This is the simplest incremental approach: Each run adds new rows to the existing table and never touches old rows. It's ideal for append-only data (e.g. logs or transactions that never change after insertion).
+
+In dbt, using append is straightforward. We configure the model as incremental and specify `incremental_strategy='append'` (for BigQuery and similar warehouses). In the SQL, we filter the source to only new records since the last load. For example, to incrementally load new transactions:
```

Comment:
```suggestion
In dbt, using append is straightforward. We configure the model as incremental and specify `incremental_strategy='append'` (for [BigQuery](/reference/resource-configs/bigquery-configs#merge-behavior-incremental-models) and similar warehouses). In the SQL, we filter the source to only new records since the last load. For example, to incrementally load new transactions:
```

Reply:
hey @Adedamola-Onabanjo , [it looks like ](https://docs.getdbt.com/docs/build/incremental-strategy#supported-incremental-strategies-by-adapter) the `append` strategy isn’t supported in BQ? ([core code here](https://github.com/dbt-labs/dbt-adapters/blob/7a568afe705d8e5884eef7416f0e7818b01aba84/dbt-bigquery/src/dbt/include/bigquery/macros/materializations/incremental.sql) -- do you have it configured and working?

Reply:
hey @Adedamola-Onabanjo , awesome thank you for clarifying! i've updated the bold bits and done with my review -- happy for you to go ahead!